### PR TITLE
committed latest notebooks

### PR DESCRIPTION
**Pull request by JakeRutherford**
# Summary

Add your summary here - keep it brief, to the point, and in plain English. [For further
information about pull requests, check out the GDS
Way](https://gds-way.cloudapps.digital/standards/pull-requests.html).

# Checklists

This pull/merge request meets the following requirements:

- [ ] code runs
- [ ] [developments are ethical][data-ethics-framework] and secure
- [ ] you have made proportionate checks that the code works correctly
- [ ] test suite passes
- [ ] developments adhere to AQA plan (see `docs/aqa/aqa_plan.md`)
- [ ] data log updated (see `docs/aqa/data_log.md`), if necessary
- [ ] assumptions, and caveats log updated (see `docs/aqa/assumptions_caveats.md`), if
  necessary
- [ ] [minimum usable documentation][agilemodeling] written in the `docs` folder

Comments have been added below around the incomplete checks.

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

### CODE REVIEW

File: notebooks/bigQuery.ipynb [hunk 1]
```diff
@@ -0,0 +1,214 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "4d79e85b",
+   "metadata": {},
+   "source": [
+    "## Big Query SQL queries used in this project"
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb]: BigQuery

File: notebooks/bigQuery.ipynb [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+   "execution_count": null,
+   "id": "dc65c551",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# import packages\n",
+    "import pandas as pd\n",
+    "import pandas_gbq\n",
+    "\n",
+    "# extract data from BigQuery\n",
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb:44]: We need to make sure users are authenticated when querying BigQuery, see (partly) Alessia's document: https://docs.google.com/document/d/1H-S0wesMFxreqbYv4g_3eegxsMnL5a_qmGRbMf1Xp60/edit#. We are using pandas_gbq to access and authenticate users. We need to use a service account credentials file to do this if we are putting this on GitHub (I'm not sure if you have one of those yet (its a json file) If not, add a message on the data labs slack channel and someone will get you one)

So we need to add something like: 
  
```
from google.oauth2 
import service_account 
import pandas_gbq
credentials = service_account.Credentials.from_service_account_file(GOOGLE_APPLICATION_CREDENTIALS,)
```

And then change the 'create dataframe' bit to: 
 
`page_path = pandas_gbq.read_gbq(sql, project_id=project_id, credentials=credentials)
` 

Therefore, we need to add the json file path key to our secrets file, for example, mine would be: 
 
`export GOOGLE_APPLICATION_CREDENTIALS="/Users/alicestephenson/gds/secrets/govuk-bigquery-analytics-credentials.json"
` 

We also need to have this somewhere in the script then:
 
`GOOGLE_APPLICATION_CREDENTIALS = os.getenv("GOOGLE_APPLICATION_CREDENTIALS") `

File: notebooks/bigQuery.ipynb [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    "WHERE\n",
+    "    _TABLE_SUFFIX BETWEEN start_date AND end_date\n",
+    "    AND hits.page.pagePath NOT LIKE \"/print%\"\n",
+    "    AND hits.type = 'PAGE'\n",
+    "    AND (SELECT value FROM hits.customDimensions WHERE index = 2) NOT IN {} \n",
+    ")\n",
+    "\n",
+    "SELECT \n",
+    "    pagePath,\n",
+    "    COUNT(sessionId) AS noOfSessions \n",
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb:136]: It doesn't impact the output here - but it's considered good practice to use DISTINCT when counting (normally at the end of the script). This way if someone skips to the end of the query they can be reassured we are deduplicating sessionIds, and it's easier for you to be sure about that as well. Note though if the sub-query table is being used as a lookup table (in a subsequent table or another script) then you might need to keep in the DISTINCT.

File: notebooks/bigQuery.ipynb [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "187147c0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# get page paths with more than 5 page hits in the last 2 weeks,\n",
+    "# ignoring irrelevant pages\n",
+    "sql = \"\"\"\n",
+    "DECLARE start_date STRING DEFAULT \"20211018\";\n",
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb:118]: For a two week period, so we don't duplicate the same days (as the day impacts which pages are visited on gov.uk), a two-week period would be 20211018 - 20211031.

File: notebooks/bigQuery.ipynb [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    "GROUP BY pagePath\n",
+    "HAVING noOfSessions >= 5\n",
+    "ORDER BY noOfSessions\n",
+    "\"\"\".format(\n",
+    "    ignore\n",
+    ")\n",
+    "\n",
+    "# create Data Frame\n",
+    "page_paths = pandas_gbq.read_gbq(sql, project_id=project_id)\n",
+    "page_paths.to_csv(\"page_paths.csv\", index=False)"
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb]: We will want the user to save it in the data folder, so this should be (as they will be in the notebooks folder): 
`page_paths.to_csv("..data/processed/page_paths.csv", index=False)
`

File: notebooks/bigQuery.ipynb [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    "FROM sessions\n",
+    "GROUP BY contentID\n",
+    "HAVING noOfSessions >= 5\n",
+    "\"\"\".format(\n",
+    "    ignore\n",
+    ")\n",
+    "\n",
+    "# create Data Frame\n",
+    "content_ids = pandas_gbq.read_gbq(sql, project_id=project_id)\n",
+    "content_ids.to_csv(\"content_ids.csv\", index=False)"
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb]: As above: 
`content_ids.to_csv("..data/processed/content_ids.csv", index=False)`

File: notebooks/bigQuery.ipynb [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1d48895a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# get all the SaB pages\n",
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb]: As querying the sabpages from BigQuery costs money, and they originally live in a csv file, I think it might be best to direct users to access the csv from google drive. In the documentation we can put 'File sab_pages.csv can be found in DataLabs shared GDrive in folder GOV.UK teams/2021-2022/Data labs/14 Network Analysis'. I spoke to Alessia about this and she agreed. This also requires authentication, so covers our back there.  So, we could just remove this query. Then in the neo4j to networkx bigquery notebook where you have: 

```
# create Data Frame 
sab_pages = pd.read_csv("sab_pages.csv")
```
 
You can change this to:
 
```
# Import libraries 
import pandas as pd 
import gspread 
from oauth2client.service_account import ServiceAccountCredentials
 
# Connect to service account
scope = ['https://spreadsheets.google.com/feeds'] 
credentials = ServiceAccountCredentials.from_json_keyfile_name(GOOGLE_APPLICATION_CREDENTIALS, scope) 
gc = gspread.authorize(credentials)
 
# Import the data from google sheets into Jupyter
spreadsheet_key = '1x3lHpUkIm-KTnWrRJU11An1J2Yb4KJT8z0QpuXOm3e0' 
book = gc.open_by_key(spreadsheet_key) 
worksheet = book.worksheet("sab_pages") 
table = worksheet.get_all_values()
 
# Convert table data into a dataframe 
pd.DataFrame(table[1:], columns=table[0])
```

Note how this also uses the secret credential `GOOGLE_APPLICATION_CREDENTIALS`, so we need to make sure we have (but with your path to the file)
`export GOOGLE_APPLICATION_CREDENTIALS="/Users/alicestephenson/gds/secrets/govuk-bigquery-analytics-credentials.json` in the .secrets file. 
 
We also need to have 
`GOOGLE_APPLICATION_CREDENTIALS = os.getenv("GOOGLE_APPLICATION_CREDENTIALS")` in the script.

File: notebooks/neo4j_to_networkx.ipynb [hunk 8]
```diff
@@ -0,0 +1,434 @@
+{
```
Comment (st-alice) [inline notebooks/neo4j_to_networkx.ipynb:1]: Just to note the page paths throughout which direct users to read or save a csv can be changed to, e.g.: 
```
nx.write_gpickle(g6, "..data/processed/structural_graph.gpickle")
```
and for example
```
page_paths = pd.read_csv("..data/processed/page_paths.csv")
```

**Review by st-alice — CHANGES_REQUESTED** [review]
Thanks JakeMsc. I haven't looked much at the community detection file as I know this work is ongoing. I've added in a couple of things but mainly to do with authentication for _1)_ BigQuery, _2)_ Google drive/sheets. 

Let me know offline if anything doesn't make sense or if you have any queries. 

I was also wondering whether we actually need to save all the files, or just the files that are used in subsequent scripts. For example, only keep and save the 1 subgraph that we will use: 

`page_paths.to_csv("..data/processed/page_paths.csv", index=False)` 

We will also need to add documentation about what these scripts are doing and why they are doing it… but we can do that when we get round to it…

File: notebooks/bigQuery.ipynb [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+   "source": [
+    "### Get SaB pages"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "765ae30c",
+   "metadata": {},
+   "source": [
+    "In order to evaluate our WUJ page finding algorithms, we require ground truth. The list of start a business (SaB) WUJ pages has been manually curated by domain experts and can be used for this purpose. Hence, this query collects the page paths of SaB pages."
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb]: I'd be tempted to remove the sab_pages query completely. We have the list of pages in GDrive that the user can access. I can't actually remember why I did it this way to begin with, but it's confusing (as in, why did I do it this way). In addition, that bigquery table no longer exists, so I am not sure how to update this code. I think the main thing is that the list of pages is in GDrive, and we have an authentication process for users to access the GDrive account.

File: notebooks/bigQuery.ipynb [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+    "GROUP BY pagePath\n",
+    "HAVING noOfSessions >= 5\n",
+    "ORDER BY noOfSessions\n",
+    "\"\"\".format(\n",
+    "    ignore\n",
+    ")\n",
+    "\n",
+    "# create Data Frame\n",
+    "page_paths = pandas_gbq.read_gbq(sql, project_id=project_id, credentials=credentials)\n",
+    "page_paths.to_csv(\"'../data/processed/page_paths.csv\", index=False)"
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb]: Remove additional `'`

File: notebooks/bigQuery.ipynb [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+    "FROM sessions\n",
+    "GROUP BY contentID\n",
+    "HAVING noOfSessions >= 5\n",
+    "\"\"\".format(\n",
+    "    ignore\n",
+    ")\n",
+    "\n",
+    "# create Data Frame\n",
+    "content_ids = pandas_gbq.read_gbq(sql, project_id=project_id, credentials=credentials)\n",
+    "content_ids.to_csv(\"'../data/processed/content_ids.csv\", index=False)"
```
Comment (st-alice) [inline notebooks/bigQuery.ipynb]: Remove additional `'`

File: notebooks/invalidNotebookFixer.ipynb [hunk 12]
```diff
@@ -0,0 +1,81 @@
+{
```
Comment (st-alice) [inline notebooks/invalidNotebookFixer.ipynb]: Not sure you want to commit this notebook?

File: notebooks/neo4j_to_networkx.ipynb [hunk 13]
```diff
@@ -0,0 +1,659 @@
...
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "55556efc",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import networkx as nx\n",
```
Comment (st-alice) [inline notebooks/neo4j_to_networkx.ipynb:20]: When trying to create the subgraphs, I need:

`from neo4j import GraphDatabase`

File: notebooks/neo4j_to_networkx.ipynb [hunk 14]
```diff
@@ -0,0 +1,659 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "510af3a7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# write the functional graph to disk\n",
+    "nx.write_gpickle(g5, \"..data/processed/functional_graph.gpickle\")"
```
Comment (st-alice) [inline notebooks/neo4j_to_networkx.ipynb]: `nx.write_gpickle(g5, "../data/processed/functional_graph.gpickle")`

File: notebooks/neo4j_to_networkx.ipynb [hunk 15]
```diff
@@ -0,0 +1,659 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8eb3fc62",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# write structural graph to disk\n",
+    "nx.write_gpickle(g6, \"..data/processed/structural_graph.gpickle\")"
```
Comment (st-alice) [inline notebooks/neo4j_to_networkx.ipynb]: `nx.write_gpickle(g6, "../data/processed/structural_graph.gpickle")`

File: notebooks/neo4j_to_networkx.ipynb [hunk 16]
```diff
@@ -0,0 +1,659 @@
...
+      ]
+     },
+     "execution_count": 19,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Import the data from google sheets\n",
+    "spreadsheet_key = '1ZIhAuP_PwITRCqrks5zag__JJ6j4gwVmmeQoKdPePI4' \n",
```
Comment (st-alice) [inline notebooks/neo4j_to_networkx.ipynb]: I think the spreadsheet key is: `1UKiLpcBAfeQZowMsA0mZAsvObAWmoUbJ6OCk7MaOSqg`

File: notebooks/neo4j_to_networkx.ipynb [hunk 17]
```diff
@@ -0,0 +1,659 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "da357ff0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# write graph to disk\n",
+    "nx.write_gpickle(g7, \"..data/processed/5_hits_graph.gpickle\")"
```
Comment (st-alice) [inline notebooks/neo4j_to_networkx.ipynb]: `nx.write_gpickle(g7, "../data/processed/5_hits_graph.gpickle")`

File: notebooks/neo4j_to_networkx.ipynb [hunk 18]
```diff
@@ -0,0 +1,659 @@
...
+      ]
+     },
+     "execution_count": 21,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Import the data from google sheets\n",
+    "spreadsheet_key = '1EvuI7AU2ei_Y6TRWxEIeKR7Yk8qCUGVSm4LuV_S6cF8' \n",
```
Comment (st-alice) [inline notebooks/neo4j_to_networkx.ipynb]: Think the key is: `13WvyoZq0gksiG7jyMGj9VnBf4URKtc6dPNoiZ72PZtk`

I've just been playing around, and it seems that it has to be a google sheets file, rather than a csv file, to be able to use this code. Guess that makes sense.

File: notebooks/neo4j_to_networkx.ipynb [hunk 19]
```diff
@@ -0,0 +1,659 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "58bfaf19",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# write graph to disk\n",
+    "nx.write_gpickle(g8, \"..data/processed/5_hits_per_contentID_graph.gpickle\")"
```
Comment (st-alice) [inline notebooks/neo4j_to_networkx.ipynb]: `nx.write_gpickle(g8, "../data/processed/5_hits_per_contentID_graph.gpickle")`

File: notebooks/randomwalks.py [hunk 20]
```diff
@@ -0,0 +1,219 @@
+import numpy as np
```
Comment (st-alice) [inline notebooks/randomwalks.py:1]: Love these functions. Great coding and documentation, really easy to follow!

**Review by st-alice — CHANGES_REQUESTED** [review]
Thanks Jake and apologies for the delay in getting this back to you. We've spoken about the structure of the repo and spitting up notebooks offline and we will add another PR to address these issues. 

Just a few comments on some of the exploratory notebooks. I was also wondering whether you want to commit the .html visualisations, or if these should go in .gitignore? I think I am under the impression that any output should go in gitignore, but happy to be told otherwise! 

I also didn't fully review the randomwalks.ipynb exploratory analyses with the knowledge graph, as the knowledge graph has been down for a few days and I wanted to get this back to you. I know that it's not paramount this is correct, as we will not be using this code in future (as we create our own graph with page view data)

File: notebooks/neo4j_to_networkx.ipynb (hunk 16 above)
Comment (JakeRutherford) [reply notebooks/neo4j_to_networkx.ipynb]: We discussed that the new plan was to get rid of the code where we import data via the google API, since it's a faff to set up. Instead, we just ask people to download the data folder from our G drive. So then they can write something like pd.read_csv('../data/page_path.csv'). So I'll do that

**Review by st-alice — APPROVED** [review]
Thanks Jake! Ready to merge :rocket:
